### ISSUE TITLE
Feature request: machine code monitor

### ISSUE THREAD

**Issue by lagomorph** [GideonZ/1541ultimate#151]
A built in monitor (even if extremely basic) that could simply examine an address and change its value would be an extremely useful enhancement. While various monitors exist in cartridge and software form, they alter the system and are not always functional in all situations.

**Comment by Zibri** [comment]
I agree. Pressing the freeze button with no cartridge loaded should go straight into a Monitor.
The one of retroreplay would suffice.

**Comment by Jusalak** [comment]
Even memory poke functionality would be a nice addition. It probably is not too difficult to implement.

**Comment by Jusalak** [comment]
A memory dump display and accompanying hex input would be a good start and far easier to implement.

**Comment by GideonZ** [comment]
Indeed. For this, the 'viewer' class should start supporting editing.
INterestingly the class is called "Editor", but it never got the edit
functions.
Sub classes could implement special operations for reading and writing, so
it would be relatively easy to edit memory.
It is just a hassle to implement this with just 'move' and 'out' functions.

**Comment by chrisgleissner** [comment]
@GideonZ I found some poke functionality in the context-sensitive menu of the U64 which you tied to a feature flag, disabled in prod. I enabled and extended it, and it seems exactly what we need as a starting point. I can peek values of the C64 RAM (C64 running, Ultimate UI in overlay) and poke them, and both sides (C64 and Ultimate UI) see the correct values.

Was there any reason why you decided to not expose the Poke functionality in prod, and should this feature also work for non-Ultimate-64 devices?

**Comment by chrisgleissner** [comment]
Very happy to finally close this one.

What started here as a request for even a basic memory monitor has grown into a full-featured Machine Code Monitor, implemented in #358 and now part of the upcoming firmware 3.15 release.

This has been a substantial piece of work for me, and I am very happy and proud to see it becoming part of the firmware after all these years. Thank you @lagomorph for starting the discussion, and everyone who contributed feedback and testing along the way. 

Also, thank you very much to @GideonZ for his help, guidance, and for trusting me with this work.

Please find the new documentation here: https://1541u-documentation.readthedocs.io/en/latest/howto/machine_code_monitor.html

I am also still working on **debugger** support in #705. This will allow for step in/out/over, 10 breakpoints, and much more. That work is now almost feature-complete, but did not quite make the cut-off for 3.15. I plan to release in in 3.16. If anyone would like to try it and provide feedback, I would be very grateful.

Closing this original feature request as completed.
